server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -109,6 +109,12 @@
         Property.Dynamic,
         Property.NodeScope
     );
+    public static final Setting<Boolean> SHARD_MOVE_PRIMARY_FIRST_SETTING = Setting.boolSetting(
```

Comment:
What is the reason for introducing a setting as opposed to always doing the "primary first" behavior?

Reply:
The primary first behavior can impact the speed of relocation, especially where some nodes only have replicas and they won't contribute to relocation initially. Hence, the default is best effort node level primary first and can be enforced across cluster level using primary first setting.